daffodil-io/src/main/scala/org/apache/daffodil/io/DirectOrBufferedDataOutputStream.scala

```diff
@@ -843,10 +860,15 @@ class DirectOrBufferedDataOutputStream private[io] (
       val nBytes = (bufferNBits / 8).toInt
       val nBytesPut = {
         if (bufOS.isFile) {
-          bufOS.flush
-          val nBitsPut = directDOS.putFile(
-            bufOS.tempFile.toPath,
-            bufferNBits.toLong, chunkSizeInBytes, finfo)
+          bufOS.close
+          val nBitsPut = try {
+            directDOS.putFile(bufOS.getFile.toPath, bufferNBits.toLong, chunkSizeInBytes, finfo)
+          } finally {
+            // Make sure we delete the file after it was put in the directDOS or
+            // if we encountered an error
+            if (bufOS.isTempFile)
+              bufOS.getFile.delete()
```

Comment:
Whatabout the case where there is an unparse error after switching to a file? In that case, this DOS will never be delivered and it won't be deleted? We need some way to delete all the temp files from all the DOS that haven't been delivered.